bundles/org.openhab.binding.folderwatcher/src/main/java/org/openhab/binding/folderwatcher/internal/handler/AzureBlobWatcherHandler.java

```diff
@@ -0,0 +1,133 @@
+
+import static org.openhab.binding.folderwatcher.internal.FolderWatcherBindingConstants.CHANNEL_NEWFILE;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.folderwatcher.internal.api.AzureActions;
+import org.openhab.binding.folderwatcher.internal.common.WatcherCommon;
+import org.openhab.binding.folderwatcher.internal.config.AzureBlobWatcherConfiguration;
+import org.openhab.core.OpenHAB;
+import org.openhab.core.io.net.http.HttpClientFactory;
+import org.openhab.core.thing.ChannelUID;
+import org.openhab.core.thing.Thing;
+import org.openhab.core.thing.ThingStatus;
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.openhab.core.types.RefreshType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link AzureBlobWatcherHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Alexandr Salamatov - Initial contribution
+ */
+@NonNullByDefault
+public class AzureBlobWatcherHandler extends BaseThingHandler {
+    private final Logger logger = LoggerFactory.getLogger(AzureBlobWatcherHandler.class);
+    private AzureBlobWatcherConfiguration config = new AzureBlobWatcherConfiguration();
+    private File currentBlobListingFile = new File(OpenHAB.getUserDataFolder() + File.separator + "AzureBlob"
+            + File.separator + thing.getUID().getAsString().replace(':', '_') + ".data");
+    private @Nullable ScheduledFuture<?> executionJob;
+    private List<String> previousBlobListing = new ArrayList<>();
+    private HttpClientFactory httpClientFactory;
+    private @Nullable AzureActions azure;
+
+    public AzureBlobWatcherHandler(Thing thing, HttpClientFactory httpClientFactory) {
+        super(thing);
+        this.httpClientFactory = httpClientFactory;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        logger.debug("Channel {} triggered with command {}", channelUID.getId(), command);
+        if (command instanceof RefreshType) {
+            refreshAzureBlobInformation();
+        }
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(AzureBlobWatcherConfiguration.class);
+
+        if (config.azureAnonymous) {
+            azure = new AzureActions(httpClientFactory, config.azureAccountName, config.azureContainerName);
+        } else {
+            azure = new AzureActions(httpClientFactory, config.azureAccountName, config.azureContainerName,
+                    config.azureAccessKey);
+        }
+
+        try {
+            previousBlobListing = WatcherCommon.initStorage(currentBlobListingFile,
```

Comment:
This method needs to return fast and it looks like this initStorage is blocking. If so it should run in a async thread.

Reply:
@lsiepel The idea was that if storage (in this case it's a file) can't be created, then thing should fail initialization phase. But if I run it in the thread, then thing will be initiated before storage is created. Would you suggest to do storage init in a different function?

Reply:
Both can still be achieved. A common procedure is like this:

1. Validat the configuration 
2. set the thing state to unknown
3. Start an scheduledtask that initilizes/authenticates etc the connection, form within that task:
 - Set the thing state accordingly
 - Schedule the `refreshAzureBlobInformation` task if all is ok.

Reply:
@lsiepel Thanks, I pushed changes, please have  a look